### DISCUSSION TITLE
How to cache StreamReponse or make a cache middleware

### DISCUSSION THREAD

**Discussion by xwzp** [fastapi/fastapi#8454] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Request
app = FastAPI()

@app.get("/test")
async def test():
    return {"hello": "cache"}

from cashews import cache
cache.setup(f"mem://?check_interval={settings.cache.check_interval}&size={settings.cache.size}")
@app.middleware("http")
async def cache_middleware(request: Request, call_next):
    key = request.scope['method'] + request.scope["path"] + str(request.scope["query_string"])
    cached_data = await cache.get(key)
    
    if cached_data is None:
        response = await call_next(request)
        await cache.set(key, response, expire=10)

        return response
    return cached_data
```

### Description

I'd like to make a general cache middleware and wrote code above. 
To my surprise, `call_next` returns a StreamResponse. I don't know exactly what it is and cached it, then program raises a expcetion tells me `h11._util.LocalProtocolError: Too little data for declared Content-Length`.

So, I wonder how to cache response correctly

### Operating System

Windows, macOS

### FastAPI Version

0.75.0

### Python Version

Python 3.8.10

**Comment by xwzp** [comment]
I found a solution to do that, but I don't know if it's a good practice. Does it `reduce performance` ? Because I convert `StreamingResponse` into normal `Response` object
Reduce performance
```Python
async def make_cached_data(response: StreamingResponse):
    content = b""
    async for chunk in response.body_iterator:
        content += chunk

    if response.media_type == 'application/json':
        response_cls = ORJSONResponse
    else:
        response_cls = Response

    return response_cls(
        status_code=response.status_code,
        content=content,
        headers=response.headers,
        media_type=response.media_type,
        background=response.background,
    )

from cashews import cache
cache.setup(f"mem://?check_interval={settings.cache.check_interval}&size={settings.cache.size}")
@app.middleware("http")
async def cache_middleware(request: Request, call_next):
    key = request.scope['method'] + request.scope["path"] + str(request.scope["query_string"])
    cached_data = await cache.get(key)
    
    # missing cache
    if cached_data is None:
        response = await call_next(request)
        
        # only cache when status code == 200
        if response.status_code == 200:
            # 写入缓存
            response = await make_cached_data(response)
            await cache.set(key, response, expire=10)
    
        return response

    # Hit cache
    return cached_data
```

**Comment by xwzp** [comment]
I looked into souce code and found my solution seems fine basically. There could be memory problem when endpoint function return a big file.

If your endpoint function return a dict, text or something not very large, fastapi always wrap all data into StreamingResponse. So it's basically fine to read data from StreamingResponse.
But if your endpoint function yield part of a large data from somthing like file, it could make memory stressful

Here are key codes for fastapi how implement middleware
```Python
            async def coro() -> None:
                nonlocal app_exc

                async with send_stream:
                    try:
                        # By calling self.app, fastapi sends message like {"type": "http.response.start"...}, 
                        # then send actual your endpoint function returns
                        await self.app(scope, request.receive, send_stream.send)
                    except Exception as exc:
                        app_exc = exc
            
            task_group.start_soon(coro)

            try:
               # Block util self.app send  {"type": "http.response.start"...}, 
                message = await recv_stream.receive()
            except anyio.EndOfStream:
                if app_exc is not None:
                    raise app_exc
                raise RuntimeError("No response returned.")
            assert message["type"] == "http.response.start"

            async def body_stream() -> typing.AsyncGenerator[bytes, None]:
                async with recv_stream:
                   # If your endpoint use return not yield, here just run once, and message is data returned by your endpoint function
                    async for message in recv_stream:
                        assert message["type"] == "http.response.body"
                        yield message.get("body", b"")

            response = StreamingResponse(
                status_code=message["status"], content=body_stream()
            )
            response.raw_headers = message["headers"]
            return response
```

**Comment by jonatasoli** [comment]
In your case with big files is better cache with database how redis ? I think cache in memory is not good way.

**Comment by xwzp** [comment]
> In your case with big files is better cache with database how redis ? I think cache in memory is not good way.

yeah, you are right. It's not a good idea to cache big file in memory. This is only suitable for small data like json

**Comment by acidjunk** [comment]
This: #4685 contains an example of an async cahce that works with a decorator on a endpoint.

**Comment by jonatasoli** [comment]
I misunderstood I'm sorry. I saw your PR and the tests is not pass yet.

**Comment by YuriiMotov** [comment]
In general it's not a good idea to cache streaming responses
